File: packages/contracts-python/anvilkit_contracts/identity_conformance.py
"""M4 identity-vector result emitter for the native Python adapter."""

from __future__ import annotations

import argparse
import base64
import hashlib
import json
import platform
from copy import deepcopy
from pathlib import Path
from typing import Any

from .canonicalizer import canonicalize
from .identity import IdentityError, component, contract_bom


def _digest(raw: bytes) -> str:
    return f"sha256:{hashlib.sha256(raw).hexdigest()}"


def _raw(value: Any) -> bytes:
    return json.dumps(value, ensure_ascii=False, separators=(",", ":")).encode("utf-8")


def _produced(canonical: bytes) -> dict[str, Any]:
    return {
        "status": "produced",
        "bytesBase64": base64.b64encode(canonical).decode("ascii"),
        "digest": _digest(canonical),
    }


def _base(case_id: str, corpus_digest: str, corpus_bytes: int) -> dict[str, Any]:
    return {
        "caseId": case_id,
        "inputDigest": corpus_digest,
        "inputBytes": corpus_bytes,
        "parseOutcome": "accepted",
    }


def _allowed(repository_root: Path) -> set[str]:
    source = json.loads((repository_root / "contracts/registries/v1/registry-set.json").read_bytes())
    registry = next(item for item in source["registries"] if item["registryId"] == "identity-purpose")
    return {entry["wireValue"] for entry in registry["entries"]}


def generate_identity(repository_root: Path) -> dict[str, Any]:
    if platform.python_version() != "3.12.7":
        raise RuntimeError(f"expected Python 3.12.7, got {platform.python_version()}")
    repository_root = repository_root.resolve()
    corpus_path = repository_root / "contracts/governance/m3/identity-cases.json"
    corpus_bytes = corpus_path.read_bytes()
    corpus = json.loads(corpus_bytes)
    if corpus["corpusVersion"] != 1 or len(corpus["componentCases"]) != 9 or len(corpus["bomCases"]) != 3:
        raise ValueError("expected identity corpus v1 with nine component and three BOM cases")
    corpus_digest = _digest(corpus_bytes)
    cases: list[dict[str, Any]] = []
    allowed = _allowed(repository_root)
    for test_case in corpus["componentCases"]:
        item = _base(test_case["id"], corpus_digest, len(corpus_bytes))
        try:
            canonical, calculated = component(test_case["value"], test_case["purpose"], test_case["mediaType"], allowed)
            if "expectedCode" in test_case or calculated != test_case["expectedDigest"]:
                raise AssertionError(f"{test_case['id']}: component expectation differs")
            item.update({
                "valid": True,
                "findings": [],
                "canonicalization": _produced(canonical),
                "componentDigest": calculated,
                "rootBomDigest": None,
                "signature": {"status": "not-applicable"},
            })
        except IdentityError as error:
            if error.code != test_case.get("expectedCode"):
                raise
            canonical = canonicalize(_raw(test_case["value"]))
            item.update({
                "valid": False,
                "findings": [{"code": error.code, "instancePath": "/", "schemaPath": "/profile/componentIdentity"}],
                "canonicalization": _produced(canonical),
                "componentDigest": None,
                "rootBomDigest": None,
                "signature": {"status": "not-applicable"},
            })
        cases.append(item)
    raw_bom_cases = {item["id"]: item for item in corpus["bomCases"]}

    def resolve(test_case: dict[str, Any]) -> Any:
        if "value" in test_case:
            return deepcopy(test_case["value"])
        source = raw_bom_cases.get(test_case.get("copyOf"))
        if source is None:
            raise ValueError(f"{test_case['id']}: invalid copyOf")
        value = resolve(source)
        if isinstance(value, dict) and "declaredDigest" in test_case:
            value["digest"] = test_case["declaredDigest"]
        return value

    for test_case in corpus["bomCases"]:
        item = _base(test_case["id"], corpus_digest, len(corpus_bytes))
        try:
            canonical, calculated, verified = contract_bom(resolve(test_case))
            if ("expectedCode" in test_case or calculated != test_case["expectedDigest"]
                    or verified != test_case["expectedVerification"]):
                raise AssertionError(f"{test_case['id']}: BOM expectation differs")
            if "expectedCanonicalWithoutDigest" in test_case and canonical.decode() != test_case["expectedCanonicalWithoutDigest"]:
                raise AssertionError(f"{test_case['id']}: canonical BOM differs")
            item.update({
                "valid": verified,
                "findings": [] if verified else [{"code": "BOM_DIGEST_MISMATCH", "instancePath": "/digest", "schemaPath": "/profile/contractBomIdentity"}],
                "canonicalization": _produced(canonical),
                "componentDigest": None,
                "rootBomDigest": calculated,
                "signature": {"status": "not-applicable"},
            })
        except IdentityError as error:
            if error.code != test_case.get("expectedCode"):
                raise
            item.update({
                "valid": False,
                "findings": [{"code": error.code, "instancePath": "/digest", "schemaPath": "/profile/contractBomIdentity"}],
                "canonicalization": {"status": "rejected", "code": error.code},
                "componentDigest": None,
                "rootBomDigest": None,
                "signature": {"status": "not-applicable"},
            })
        cases.append(item)
    cases.sort(key=lambda item: item["caseId"].encode("utf-8"))
    adapter_bytes = b"".join(
        (Path(__file__).parent / name).read_bytes()
        for name in ("identity.py", "canonicalizer.py", "identity_conformance.py")
    )
    return {
        "resultVersion": 1,
        "fixtureManifestDigest": corpus_digest,
        "language": "python",
        "implementation": {
            "adapterId": "anvilkit-python-identity",
            "adapterVersion": "0.1.0",
            "runtime": "python",
            "runtimeVersion": "3.12.7",
            "adapterDigest": _digest(adapter_bytes),
        },
        "cases": cases,
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--repository-root", type=Path, default=Path(__file__).resolve().parents[3])
    args = parser.parse_args()
    print(json.dumps(generate_identity(args.repository_root), indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: packages/contracts-java/src/main/java/dev/anvilkit/contracts/IdentityConformance.java
package dev.anvilkit.contracts;

import tools.jackson.databind.JsonNode;
import tools.jackson.databind.json.JsonMapper;
import tools.jackson.databind.node.ArrayNode;
import tools.jackson.databind.node.ObjectNode;

import java.io.ByteArrayOutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.HexFormat;
import java.util.List;
import java.util.Map;
import java.util.Set;

/** Emits the 12-case native Java identity conformance result. */
public final class IdentityConformance {
    private static final JsonMapper MAPPER = JsonMapper.builder().build();

    private IdentityConformance() {}

    private static String digest(byte[] raw) throws Exception {
        return "sha256:" + HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(raw));
    }

    private static ObjectNode base(String id, String corpusDigest, int corpusBytes) {
        ObjectNode item = MAPPER.createObjectNode();
        item.put("caseId", id);
        item.put("inputDigest", corpusDigest);
        item.put("inputBytes", corpusBytes);
        item.put("parseOutcome", "accepted");
        return item;
    }

    private static ObjectNode produced(byte[] canonical) throws Exception {
        ObjectNode value = MAPPER.createObjectNode();
        value.put("status", "produced");
        value.put("bytesBase64", Base64.getEncoder().encodeToString(canonical));
        value.put("digest", digest(canonical));
        return value;
    }

    private static ArrayNode findings(String code, String instancePath, String schemaPath) {
        ArrayNode findings = MAPPER.createArrayNode();
        ObjectNode finding = findings.addObject();
        finding.put("code", code);
        finding.put("instancePath", instancePath);
        finding.put("schemaPath", schemaPath);
        return findings;
    }

    private static Set<String> allowedPurposes(Path repositoryRoot) throws Exception {
        JsonNode source = MAPPER.readTree(Files.readAllBytes(repositoryRoot.resolve("contracts/registries/v1/registry-set.json")));
        Set<String> allowed = new HashSet<>();
        for (JsonNode registry : source.get("registries")) {
            if (registry.get("registryId").asString().equals("identity-purpose")) {
                for (JsonNode entry : registry.get("entries")) allowed.add(entry.get("wireValue").asString());
            }
        }
        return allowed;
    }

    private static JsonNode resolveBom(JsonNode testCase, Map<String, JsonNode> cases) {
        if (testCase.has("value")) return testCase.get("value").deepCopy();
        JsonNode source = cases.get(testCase.get("copyOf").asString());
        if (source == null) throw new IllegalStateException(testCase.get("id").asString() + ": invalid copyOf");
        JsonNode value = resolveBom(source, cases);
        if (value instanceof ObjectNode object && testCase.has("declaredDigest")) {
            object.put("digest", testCase.get("declaredDigest").asString());
        }
        return value;
    }

    public static ObjectNode generate(Path repositoryRoot) throws Exception {
        if (Runtime.version().feature() != 17) throw new IllegalStateException("expected JDK 17, got " + Runtime.version());
        repositoryRoot = repositoryRoot.toAbsolutePath().normalize();
        Path corpusPath = repositoryRoot.resolve("contracts/governance/m3/identity-cases.json");
        byte[] corpusBytes = Files.readAllBytes(corpusPath);
        JsonNode corpus = MAPPER.readTree(corpusBytes);
        if (corpus.get("corpusVersion").asInt() != 1 || corpus.get("componentCases").size() != 9 || corpus.get("bomCases").size() != 3) {
            throw new IllegalStateException("expected identity corpus v1 with nine component and three BOM cases");
        }
        String corpusDigest = digest(corpusBytes);
        Set<String> allowed = allowedPurposes(repositoryRoot);
        List<ObjectNode> cases = new ArrayList<>();
        for (JsonNode testCase : corpus.get("componentCases")) {
            ObjectNode item = base(testCase.get("id").asString(), corpusDigest, corpusBytes.length);
            try {
                var actual = ContractIdentity.component(testCase.get("value"), testCase.get("purpose").asString(), testCase.get("mediaType").asString(), allowed);
                if (testCase.has("expectedCode") || !actual.digest().equals(testCase.get("expectedDigest").asString())) {
                    throw new IllegalStateException(testCase.get("id").asString() + ": component expectation differs");
                }
                item.put("valid", true);
                item.set("findings", MAPPER.createArrayNode());
                item.set("canonicalization", produced(actual.canonical()));
                item.put("componentDigest", actual.digest());
                item.putNull("rootBomDigest");
            } catch (ContractIdentity.IdentityException error) {
                if (!testCase.has("expectedCode") || !error.code().equals(testCase.get("expectedCode").asString())) throw error;
                var canonical = ContractIdentity.component(testCase.get("value"), "schema", "application/json", Set.of("schema")).canonical();
                item.put("valid", false);
                item.set("findings", findings(error.code(), "/", "/profile/componentIdentity"));
                item.set("canonicalization", produced(canonical));
                item.putNull("componentDigest");
                item.putNull("rootBomDigest");
            }
            item.putObject("signature").put("status", "not-applicable");
            cases.add(item);
        }
        Map<String, JsonNode> rawBomCases = new HashMap<>();
        for (JsonNode testCase : corpus.get("bomCases")) rawBomCases.put(testCase.get("id").asString(), testCase);
        for (JsonNode testCase : corpus.get("bomCases")) {
            ObjectNode item = base(testCase.get("id").asString(), corpusDigest, corpusBytes.length);
            try {
                var actual = ContractIdentity.contractBom(resolveBom(testCase, rawBomCases));
                if (testCase.has("expectedCode") || !actual.digest().equals(testCase.get("expectedDigest").asString())
                        || actual.verified() != testCase.get("expectedVerification").asBoolean()) {
                    throw new IllegalStateException(testCase.get("id").asString() + ": BOM expectation differs");
                }
                if (testCase.has("expectedCanonicalWithoutDigest")
                        && !new String(actual.canonical(), java.nio.charset.StandardCharsets.UTF_8).equals(testCase.get("expectedCanonicalWithoutDigest").asString())) {
                    throw new IllegalStateException(testCase.get("id").asString() + ": canonical BOM differs");
                }
                item.put("valid", actual.verified());
                item.set("findings", actual.verified() ? MAPPER.createArrayNode() : findings("BOM_DIGEST_MISMATCH", "/digest", "/profile/contractBomIdentity"));
                item.set("canonicalization", produced(actual.canonical()));
                item.putNull("componentDigest");
                item.put("rootBomDigest", actual.digest());
            } catch (ContractIdentity.IdentityException error) {
                if (!testCase.has("expectedCode") || !error.code().equals(testCase.get("expectedCode").asString())) throw error;
                item.put("valid", false);
                item.set("findings", findings(error.code(), "/digest", "/profile/contractBomIdentity"));
                ObjectNode rejected = item.putObject("canonicalization");
                rejected.put("status", "rejected");
                rejected.put("code", error.code());
                item.putNull("componentDigest");
                item.putNull("rootBomDigest");
            }
            item.putObject("signature").put("status", "not-applicable");
            cases.add(item);
        }
        cases.sort(Comparator.comparing(item -> item.get("caseId").asString()));
        ByteArrayOutputStream adapterBytes = new ByteArrayOutputStream();
        for (String name : List.of("ContractIdentity.java", "JcsCanonicalizer.java", "IdentityConformance.java")) {
            adapterBytes.write(Files.readAllBytes(repositoryRoot.resolve("packages/contracts-java/src/main/java/dev/anvilkit/contracts").resolve(name)));
        }
        ObjectNode output = MAPPER.createObjectNode();
        output.put("resultVersion", 1);
        output.put("fixtureManifestDigest", corpusDigest);
        output.put("language", "java");
        ObjectNode implementation = output.putObject("implementation");
        implementation.put("adapterId", "anvilkit-java-identity");
        implementation.put("adapterVersion", "0.1.0");
        implementation.put("runtime", "jdk");
        implementation.put("runtimeVersion", "17");
        implementation.put("adapterDigest", digest(adapterBytes.toByteArray()));
        ArrayNode outputCases = output.putArray("cases");
        for (ObjectNode item : cases) outputCases.add(item);
        return output;
    }

    public static void main(String[] args) throws Exception {
        Path repositoryRoot = Path.of("../..");
        for (int index = 0; index + 1 < args.length; index++) {
            if (args[index].equals("--repository-root")) repositoryRoot = Path.of(args[index + 1]);
        }
        System.out.println(MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(generate(repositoryRoot)));
    }
}
